Update Certificate Validation with more clarity and tests (#4165)
projectcontour/contour: site/content/docs/main/config/tls-termination.md

When TLS is enabled for a virtual host, any request to the insecure port is redirected to the secure interface with a 301 redirect.
Specific routes can be configured to override this behavior and handle insecure requests by enabling the `spec.routes.permitInsecure` parameter on a Route.

The TLS secret must:
- be a Secret of type `kubernetes.io/tls`. This means that it must contain keys named `tls.crt` and `tls.key` that contain the certificate and private key to use for TLS, in PEM format.

The TLS secret may also:

> - add any additional certificates required for validation into the `tls.crt` PEM bundle. If this is the case, the serving certificate must be the first certificate in the bundle.

Reviewer: Shouldn't the first bullet be removed? The additional certs in `tls.crt` should be **only** the chain CA certs which are there to support the peer when it is validating of our serving cert. Sending other certs would be redundant.

Author: Yep, I missed this, thanks.

When TLS is enabled for a virtual host, any request to the insecure port is redirected to the secure interface with a 301 redirect.
Specific routes can be configured to override this behavior and handle insecure requests by enabling the `spec.routes.permitInsecure` parameter on a Route.

The TLS secret must:
- be a Secret of type `kubernetes.io/tls`. This means that it must contain keys named `tls.crt` and `tls.key` that contain the certificate and private key to use for TLS, in PEM format.

The TLS secret may also:

> - add any additional certificates required for validation into the `tls.crt` PEM bundle. If this is the case, the serving certificate must be the first certificate in the bundle.
> - contain a `ca.crt` key that contains a certificate bundle to use for CA validation.
> 
> We recommend using only one of these options.
> 
> Practically, this means that, if you have been issued a certificate bundle from a CA that contains the certificate plus intermediate certificates, you can put the contents of that bundle into `tls.crt`, and everything will work.
> If you have a separate file with an intermediate certificate chain, you may put those details into a `ca.crt` key, and everything will work.

Reviewer: Maybe consider reformulating, something along the lines of: 

```suggestion
- add any chain CA certificates required for validation into the `tls.crt` PEM bundle. If this is the case, the serving certificate must be the first certificate in the bundle and the intermediate CA certificates must be appended in issuing order.
```

I think everything else could be removed, because AFAIK `ca.crt` is not used currently and I would not change that. I believe it is correct behavior. It is pretty popular convention to have chain certs appended and not as separate CA bundle. Envoy also expects the server certificate and its chain in field `CertificateChain` in the SDS secret ([doc link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto.html?highlight=tls#extensions-transport-sockets-tls-v3-tlscertificate)) and currently we just map `tls.crt` contents into `PrivateKey` and `CertificateChain` fields. We do not additionally append the contents of `ca.crt` to CertificateChain.
